File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.length();
        int len = 0, start = 0;
        vector<vector<int>> dp(n+1,vector<int> (n+1,0));
        for(int i=0;i<n;i++){
            dp[i][i] = 1;
            len = 1;
            start = i;
        }
        for(int i=0;i<n-1;i++){
            if(s[i] == s[i+1]){
                dp[i][i+1] = 1;
                len = 2;
                start = i;
            }
        }
        
        for(int j=2;j<n;j++){
            for(int i=0;i<n-j+1;i++){
                if(dp[i+1][j+i-1] and s[i] == s[j+i]){
                    dp[i][j+i] = 1;
                    len = j+1;
                    start = i;
                }
            }
        }
        // for(int i=0;i<n;i++){
        //     for(int j=0;j<n;j++){
        //         cout<<dp[i][j]<<" ";
        //     }
        //     cout<<endl;
        // }
        return s.substr(start,len);
    }
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    let dp = [];
    let start = 0, end = 0, max = 0;;
    for(let i=0;i<s.length;i++){
        const tempArr = [];
        for(let j=0;j<s.length;j++){
            if(i === j || i === j+1){
                tempArr.push(1)
            }
            else {
                tempArr.push(0);
            }
        }
        dp.push(tempArr);
    }
    let k = 0;
    while(k < s.length - 1){
        let i = 0, j = k+1;
        while(j < s.length){
            if(s[i] === s[j]){
                if(j-i === 1){
                    dp[i][j] = 2;
                }
                else{
                    dp[i][j] = dp[i+1][j-1] ? dp[i+1][j-1] + 2 : 0;
                }
                if(max < dp[i][j]){
                    max = dp[i][j];
                    start = i;
                    end = j;
                }
            }
            j++;
            i++;
        }
        k++;
    }
    console.log(dp);
    return s.slice(start, end+1);
};